x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,203 @@
+                        SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                    );
+                    if (Objects.equals(indexId, otherIndexId)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    static QueryBuilder buildDeleteByQuery(int numberOfShards, String snapshotUuid, String indexUuid) {
+        final Set<String> paths = new HashSet<>(numberOfShards);
+        for (int shard = 0; shard < numberOfShards; shard++) {
+            paths.add(String.join("/", snapshotUuid, indexUuid, String.valueOf(shard)));
+        }
+        return QueryBuilders.termsQuery("blob.path", paths);
+    }
+
+    private class MaintenanceTask extends AbstractRunnable {
+
+        private final ClusterChangedEvent event;
+
+        MaintenanceTask(ClusterChangedEvent event) {
+            assert event.indicesDeleted().isEmpty() == false;
+            this.event = Objects.requireNonNull(event);
+        }
+
+        @Override
+        protected void doRun() {
+            final ClusterState state = event.state();
+            for (Index deletedIndex : event.indicesDeleted()) {
+                final IndexMetadata indexMetadata = event.previousState().metadata().index(deletedIndex);
+                if (indexMetadata != null) {
+                    final Settings indexSetting = indexMetadata.getSettings();
+                    if (SearchableSnapshotsSettings.isSearchableSnapshotStore(indexSetting)) {
+                        assert state.metadata().hasIndex(deletedIndex) == false;
+
+                        final SnapshotId snapshotId = new SnapshotId(
+                            SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSetting),
+                            SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSetting)
+                        );
+                        final IndexId indexId = new IndexId(
+                            SNAPSHOT_INDEX_NAME_SETTING.get(indexSetting),
+                            SNAPSHOT_INDEX_ID_SETTING.get(indexSetting)
+                        );
+
+                        // we should do nothing if the current cluster state contains another
+                        // searchable snapshot index that uses the same index snapshot
+                        if (hasSearchableSnapshotWith(state, snapshotId, indexId)) {
+                            logger.debug(
+                                "snapshot [{}] of index {} is in use, skipping maintenance of snapshot blob cache entries",
+                                snapshotId,
+                                indexId
+                            );
+                            continue;
+                        }
+
+                        final DeleteByQueryRequest request = new DeleteByQueryRequest(systemIndexName);
+                        request.setQuery(buildDeleteByQuery(indexMetadata.getNumberOfShards(), snapshotId.getUUID(), indexId.getId()));
```

Comment:
Do we need to ask delete-by-query to refresh to ensure it can see the newest entries, at least on the first request?

Reply:
For simplification I added the refresh for the 1st delete-by-query